File: backend/api/schemas/stock.py
"""
Pydantic request/response schemas for stock-related endpoints.
All API contracts are defined here — no loose dicts as public contracts.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Envelope
# ---------------------------------------------------------------------------

class SuccessResponse(BaseModel):
    success: bool = True
    data: Any = None
    message: str = "操作成功"


class ErrorResponse(BaseModel):
    success: bool = False
    error: str
    code: Optional[str] = None


# ---------------------------------------------------------------------------
# Stock info
# ---------------------------------------------------------------------------

class StockInfo(BaseModel):
    symbol: str
    name: str
    price: Optional[float] = None
    change: Optional[float] = None
    changePercent: Optional[float] = None
    volume: Optional[int] = None
    marketCap: Optional[int] = None
    sector: str = ""
    industry: str = ""
    market: str = ""
    currency: str = ""
    lastUpdated: str = ""
    available: bool = True
    source: Optional[str] = None
    as_of: Optional[str] = None
    data_version: Optional[str] = None
    is_stale: bool = False
    reason: Optional[str] = None


# ---------------------------------------------------------------------------
# History
# ---------------------------------------------------------------------------

class OHLCVItem(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int
    amount: Optional[float] = None
    interval: Optional[str] = None
    adjust_type: Optional[str] = None
    source: Optional[str] = None


class StockHistoryResponse(BaseModel):
    symbol: str
    period: str
    interval: str
    data: List[OHLCVItem]
    indicators: Optional[Dict[str, Any]] = None


# ---------------------------------------------------------------------------
# Financials
# ---------------------------------------------------------------------------

class FinancialItem(BaseModel):
    symbol: str
    year: int
    quarter: int
    revenue: Optional[float] = None
    net_profit: Optional[float] = None
    gross_margin: Optional[float] = None
    net_margin: Optional[float] = None
    operating_cash_flow: Optional[float] = None
    eps: Optional[float] = None
    roe: Optional[float] = None
    revenue_yoy: Optional[float] = None
    profit_yoy: Optional[float] = None
    price_yoy: Optional[float] = None
    report_date: Optional[str] = None
    announced_at: Optional[str] = None
    period_type: Optional[str] = None
    currency: Optional[str] = None
    source: Optional[str] = None
    data_version: Optional[str] = None
    is_restated: bool = False
    total_assets: Optional[float] = None
    total_liabilities: Optional[float] = None
    shareholder_equity: Optional[float] = None
    cash_and_equivalents: Optional[float] = None
    interest_bearing_debt: Optional[float] = None
    accounts_receivable: Optional[float] = None
    inventory: Optional[float] = None
    gross_profit: Optional[float] = None
    operating_profit: Optional[float] = None
    investing_cash_flow: Optional[float] = None
    financing_cash_flow: Optional[float] = None
    free_cash_flow: Optional[float] = None
    capex: Optional[float] = None


# ---------------------------------------------------------------------------
# Search
# ---------------------------------------------------------------------------

class StockSearchResult(BaseModel):
    symbol: str
    name: str
    market: str = ""


# ---------------------------------------------------------------------------
# Research report
# ---------------------------------------------------------------------------

class EPSForecast(BaseModel):
    eps: float
    pe: Optional[float] = None


class ResearchReport(BaseModel):
    title: str = ""
    rating: str = ""
    institution: str = ""
    date: str = ""
    industry: str = ""
    pdf_url: str = ""
    eps_forecast: Dict[str, EPSForecast] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# AI analysis
# ---------------------------------------------------------------------------

class AIAnalysisResponse(BaseModel):
    symbol: str
    name: str
    current_price: Optional[float] = None
    industry: Optional[str] = None
    analysis: str = ""
    sources: str = ""
    error: Optional[str] = None

File: frontend/src/types/stock.ts
// Shared stock-domain types.
// Keep in sync with backend/api/schemas/stock.py.

export interface StockInfo {
  symbol: string
  name: string
  price: number | null
  change: number | null
  changePercent: number | null
  volume: number | null
  marketCap: number | null
  sector: string
  industry: string
  market: string
  currency: string
  lastUpdated: string
  available?: boolean
  source?: string | null
  as_of?: string | null
  data_version?: string | null
  is_stale?: boolean
  reason?: string | null
}

export interface OHLCVItem {
  date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
  amount?: number | null
  interval?: string | null
  adjust_type?: string | null
  source?: string | null
}

export interface StockHistory {
  symbol: string
  period: string
  interval: string
  data: OHLCVItem[]
  indicators?: Record<string, unknown>
}

export interface FinancialItem {
  symbol: string
  year: number
  quarter: number
  revenue: number | null
  net_profit: number | null
  gross_margin: number | null
  net_margin: number | null
  operating_cash_flow: number | null
  eps: number | null
  roe: number | null
  revenue_yoy?: number | null
  profit_yoy?: number | null
  price_yoy?: number | null
  report_date?: string | null
  announced_at?: string | null
  period_type?: string | null
  currency?: string | null
  source?: string | null
  data_version?: string | null
  is_restated?: boolean
  total_assets?: number | null
  total_liabilities?: number | null
  shareholder_equity?: number | null
  cash_and_equivalents?: number | null
  interest_bearing_debt?: number | null
  accounts_receivable?: number | null
  inventory?: number | null
  gross_profit?: number | null
  operating_profit?: number | null
  investing_cash_flow?: number | null
  financing_cash_flow?: number | null
  free_cash_flow?: number | null
  capex?: number | null
}

export interface StockSearchResult {
  symbol: string
  name: string
  market?: string
}

export interface EPSForecast {
  eps: number
  pe?: number | null
}

export interface ResearchReport {
  title: string
  rating: string
  institution: string
  date: string
  industry: string
  pdf_url: string
  eps_forecast: Record<string, EPSForecast>
}

export interface AIAnalysisResult {
  symbol: string
  name: string
  current_price?: number
  industry?: string
  analysis: string
  sources: string
  error?: string
}

export interface ResearchSummaryResult {
  symbol: string
  name: string
  current_price?: number
  report_count?: number
  summary: string
  reports: ResearchReport[]
  sources: string
  error?: string
}
